rust/lance-index/src/vector/hnsw/builder.rs

```diff
@@ -87,144 +111,470 @@ impl HnswBuildParams {
-        Self { inner }
-    }
+    // pub fn with_params(
+    //     distance_type: DistanceType,
+    //     params: HnswBuildParams,
+    //     vectors: Option<Arc<dyn VectorStorage>>,
+    // ) -> Self {
+    //     let inner = Arc::new(HNSWBuilderInner::with_params(
+    //         distance_type,
+    //         params,
+    //         vectors,
+    //     ));
+    //     Self { inner }
+    // }
 
     /// Build the graph, with the already provided `VectorStorage` as backing storage for HNSW graph.
-    pub async fn build(&mut self) -> Result<HNSW> {
+    pub async fn build_with_storage(
+        distance_type: DistanceType,
+        params: HnswBuildParams,
+        storage: Arc<dyn VectorStorage>,
+    ) -> Result<Self> {
+        let inner = HNSWBuilderInner::with_params(distance_type, params, storage.as_ref());
+        let hnsw = Self {
+            inner: Arc::new(inner),
+        };
+
         log::info!(
             "Building HNSW graph: num={}, metric_type={}, max_levels={}, m={}, ef_construction={}",
-            self.inner.vectors.len(),
-            self.inner.vectors.metric_type(),
-            self.inner.params.max_level,
-            self.inner.params.m,
-            self.inner.params.ef_construction
+            storage.len(),
+            hnsw.inner.distance_type,
+            hnsw.inner.params.max_level,
+            hnsw.inner.params.m,
+            hnsw.inner.params.ef_construction
         );
-
-        if self.inner.vectors.len() <= 1 {
-            return Ok(HNSW::from_builder(
-                self,
-                self.inner.entry_point,
-                self.inner.vectors.metric_type(),
-                self.inner.params.clone(),
-            ));
+        if storage.len() <= 1 {
+            return Ok(hnsw);
         }
 
-        let parallel_limit = self
+        let len = storage.len();
+        let parallel_limit = hnsw
             .inner
             .params
             .parallel_limit
-            .unwrap_or_else(num_cpus::get);
+            .unwrap_or_else(num_cpus::get)
+            .max(1);
+        log::info!("Building HNSW graph with parallel_limit={}", parallel_limit);
         let mut tasks = Vec::with_capacity(parallel_limit);
-        let chunk_size = (self.inner.vectors.len() - 1).div_ceil(parallel_limit);
-        for chunk in &(1..self.inner.vectors.len()).chunks(chunk_size) {
+        let chunk_size = (storage.len() - 1).div_ceil(parallel_limit);
+        for chunk in &(1..storage.len()).chunks(chunk_size) {
             let chunk = chunk.collect_vec();
-            let inner = self.inner.clone();
+            let inner = hnsw.inner.clone();
+            let storage = storage.clone();
+            let visited_generator = VisitedGenerator::new(len);
             tasks.push(spawn_cpu(move || {
-                for node in chunk {
-                    inner.insert(node as u32)?;
+                for node in kdam::tqdm!(chunk.into_iter()) {
```

Comment:
Remove kdam use here. It doesn't render correctly anyways (due to the multi tasking).